Svc

LindaYu17 opened the pull request.

LindaYu17 committed:
interface: fix typo

Signed-off-by: Yu, Linda <linda.yu@intel.com>

LindaYu17 committed:
encoder: m_temporalID defined in base class should be used

There are two temporal id members for VaapiEncPictureH264 which are
m_temporalId and m_temporalID, only one should be used.

Signed-off-by: Yu, Linda <linda.yu@intel.com>

xuguangxin commented:
@wudping , please help review this.
Could we add some test case in daily regression to capture svct issues?

wudping commented on v4l2/v4l2_encode.cpp:
```diff
@@ -58,6 +58,14 @@ bool V4l2Encoder::start()
+        VideoTemporalLayers layers;
+        layers.numLayersMinus1 = 1; // It is two layers actually
+        layers.bitRate[0] = 3000000;
+        layers.bitRate[1] = 4000000;
```
The highest layer's bitrate should been set via "VideoParamsCommon.rcParams.bitRate"; So this line should be replaced with:
VideoParamsCommon.rcParams.bitRate = 4000000;

LindaYu17 commented:
@wudping updated!

xuguangxin commented on v4l2/v4l2_encode.cpp:
```diff
@@ -195,6 +203,16 @@ bool V4l2Encoder::giveOutputBuffer(struct v4l2_buffer *dqbuf)
         dqbuf->flags = V4L2_BUF_FLAG_KEYFRAME;
 
+
+    printf("outputBuffer->temporalId = %d\n", outputBuffer->temporalID);
```
please remove this print?

wudping commented:
+1

wudping commented:
+1

xuguangxin commented on v4l2/v4l2_encode.cpp:
```diff
@@ -58,6 +58,14 @@ bool V4l2Encoder::start()
+    if (m_svct) {
+        VideoTemporalLayers layers;
+        layers.numLayersMinus1 = 1; // It is two layers actually
+        layers.bitRate[0] = 3000000;
```
This hard coded bitrate may introduce issue on dynamic bitrate changes. Could we change logical like this:
if total birate is a, the zero layer bitrate is a/2 etc.

LindaYu17 replied:
do you mean this?
layers.bitRate[0] = m_videoParams.rcParams.bitRate / 2;
layers.bitRate[1] = m_videoParams.rcParams.bitRate;

but if we don't set m_videoParams.rcParams.bitRate, how to set the layer bitrate？

wudping replied:
In my opinion, m_videoParams.rcParams.bitRate is what we need to set here.

wudping replied:
How about set a default bitrate according to the "m_videoParams.resolution" in case m_videoParams.rcParams.bitRate == 0. As far as I know, 4 Mb/s for 1080P.

wudping commented on v4l2/v4l2_encode.cpp:
```diff
@@ -58,6 +58,15 @@ bool V4l2Encoder::start()
+        VideoTemporalLayers layers;
+        layers.numLayersMinus1 = 1; // It is two layers actually
+        layers.bitRate[0] = m_videoParams.rcParams.bitRate / 2;
+        layers.bitRate[1] = m_videoParams.rcParams.bitRate;
```
Hi @LindaYu17 , 
We can delete the line "layers.bitRate[1] = m_videoParams.rcParams.bitRate;", because the second layer is the highest layer in this case, and whose bitrate is assigned by "m_videoParams.rcParams.bitRate" in vaapiencoder_xxx.cpp file.

LindaYu17 replied:
layers.bitRate[1] = m_videoParams.rcParams.bitRate;
this assignment make the code more clear and it is not conflict with that in libyami either, so suggest to keep this line.

LindaYu17 committed:
v4l2: enable svct in v4l2

Expose svc-t enable interface in v4l2 to user.
Temporal layer number setting and bitrate setting for each layer
interface are not exposed yet.

Signed-off-by: Yu, Linda <linda.yu@intel.com>

xuguangxin commented:
@wudping , is it ok to merge this?

wudping commented:
@xuguangxin , it's okay for me.

xuguangxin commented:
@wudping this will break yami-utils since a API change, right? could you prepare a patch for yami-utils?

wudping commented:
@xuguangxin , okay, I am planing to add a patch for this update next week. But SVC-T regression will be latter.

wudping commented:
hi @xuguangxin, The commit of interface modifying the typo "VideoTemproalLayers" does not impact libyami-utils.

xuguangxin commented:
@wudping , why? this means the temporal layer never set in yami-utils?

wudping commented:
Hi @xuguangxin , 
typedef struct VideoParamsCommon {
    uint32_t size;
    VAProfile profile;
    uint8_t level;
    VideoResolution resolution;
    VideoFrameRate frameRate;
    **VideoTemproalLayers temporalLayers;**
libyami-itils sets temporalLayers via VideoParamsCommon, and yami-utils needn't to define a new variable with VideoTemproalLayers type, so we just need to modify the file VideoEncoderDefs.h in libyami.

Anyway, I am sorry for this typo. It's my fault.

xuguangxin commented:
ok thanks for explain
